### DISCUSSION TITLE
Call (custom) page action from backend

### DISCUSSION THREAD

**Discussion by studiowizjo** [filamentphp/filament#6970] [Help]
I have a custom "importFromCsv" page action which opens modal with form:

```php
class ListArticles extends ListRecords
{
    protected static string $resource = ArticleResource::class;

    public function importArticlesFromCsvAction()
    {
        return Action::make('importFromCsv')
            ->label('Import from .csv')
            ->icon('heroicon-o-upload')
            ->form($this->getImportArticlesFromCsvActionForm())
            ->color('secondary')
            ->action(function (array $data) {
                //...
            });
    }

    protected function getActions(): array
    {
        return [
            $this->importArticlesFromCsvAction(),
            Actions\CreateAction::make(),
        ];
    }
}
```

I would like to have a link/button which opens this action from any place in app. Because I think it can be very problematic to do so from every place, the accepted solution can be also to redirect user to list record page with get parameter and base on that call action.

I have tried below solutions, but none of them works:

```php
class ListArticles extends ListRecords
{
    public function mount(): void
    {
        parent::mount();

        if (auth()->user()->isAdmin() && request()->has('openBulkImport')) {
            $this->callMountedAction('importFromCsv');
            $this->mountAction('importFromCsv');
            $this->emit('openModal', 'importFromCsv');
            $this->emit('openModal', 'import-from-csv');

            $this->dispatchBrowserEvent('open-modal', [
                 'id' => 'import-from-csv',
             ]);
        }
    }
```

I know I could call it from frontend, but if it is possible from backend, I would prefer such solution.

**Comment by rmitesh** [comment]
@studiowizjo I guess you can create a trait for the same and use it in every class and merge it with the actions 

```php
trait ImportActions
{
	public function getImportActions(): array
	{
		return [
			Actions\Action::make('Import'),
		];
	}
}
```

```php
protected function getActions(): array
{
    return array_merge($this->getImportActions(), [
        Actions\CreateAction::make(),
    ]);
}
```

OR

If you are not planning to add more actions to the page, then just implement like this

```php
trait ImportActions
{
	protected function getActions(): array
	{
	    return array_merge([
	    	Actions\Action::make('Import'),
	    ], parent::getActions());
	}
}
```
and comment `getActions()` function for the resource pages.

**Reply by studiowizjo** [reply]
In my code it is indeed in trait, only here it is a little modified for a simplicity, but this is not the topic I`m asking about.

**Reply by rmitesh** [reply]
@studiowizjo could you elaborate on what you are looking for?

**Comment by pboivin** [comment]
@studiowizjo It looks like this one should work: `$this->mountAction('importFromCsv');` 
Have you tried calling it from another place? (not `mount()`)

**Reply by studiowizjo** [reply]
Yes, I did, but maybe you have some specific idea where to trigger it?

**Reply by pboivin** [reply]
Hmm... I'm starting to think that the best way is indeed from the front-end. Maybe a custom element on your page with an `x-init` attribute?
```html
<div x-data x-init="
    if ( /* check for query string parameter  */ ) {
        $wire.mountAction('importFromCsv');
    }
">
</div>
```

**Reply by studiowizjo** [reply]
Yes, I know how to do it from frontend, but I was hoping it is possible from backend, so I don`t have to make custom view just for that or use any "hacky" solutions to inject it, for example, via placeholder field.

**Comment by sprklinginfo** [comment]
@studiowizjo I am working on a similar feature. Have you figured out how to programmatically trigger the import action modal from the backend? Thanks!

**Reply by studiowizjo** [reply]
Unfortunately not =(.
